merge_all.py

'print' is not compatible with python 3 and I met with another
'd = open(os.path.join(header_path, header)).read().decode('utf8')
AttributeError: 'str' object has no attribute 'decode''. I've tried to remove the '.decode('utf8')' and seemly solved it.
Beyond this issue, I finally met with 'crow/tests/unittest.cpp:936:46: error: call of overloaded ‘move(const crow::query_string&)’ is ambiguous
         last_url_params = move(req.url_params);
                                              ^
'
build with GNU Make 4.1, g++ (GCC) 5.1.0, Python 3.4.3, boost 1.58.0 on archlinux with kernel version 4.0.5, x86_64 arch

--- a/amalgamate/merge_all.py
+++ b/amalgamate/merge_all.py
@@ -1,5 +1,6 @@
-import glob
-import os
+"""Merges all the header files."""
+from glob import glob
+from os import path as pt
 import re
 from collections import defaultdict
 import sys
@@ -10,11 +11,11 @@ if len(sys.argv) > 1:
 
 OUTPUT = 'crow_all.h'
 re_depends = re.compile('^#include "(.*)"', re.MULTILINE)
-headers = [x.rsplit('/',1)[-1] for x in glob.glob(os.path.join(header_path,'*.h'))]
-print headers
+headers = [x.rsplit('/', 1)[-1] for x in glob(pt.join(header_path, '*.h'))]
+print(headers)
 edges = defaultdict(list)
 for header in headers:
-    d = open(os.path.join(header_path, header)).read().decode('utf8')
+    d = open(pt.join(header_path, header)).read().decode('utf8')
     match = re_depends.findall(d)
     for m in match:
         # m should included before header
@@ -22,30 +23,32 @@ for header in headers:
 
 visited = defaultdict(bool)
 order = []
+
+
 def dfs(x):
+    """Ensure all header files are visited."""
     visited[x] = True
     for y in edges[x]:
         if not visited[y]:
             dfs(y)
     order.append(x)
-            
+
 for header in headers:
     if not visited[header]:
         dfs(header)
 
 order = order[::-1]
 for x in edges:
-    print x, edges[x]
+    print(x, edges[x])
 for x in edges:
     for y in edges[x]:
         assert order.index(x) < order.index(y), 'cyclic include detected'
 
-print order
+print(order)
 build = []
 for header in order:
-    d = open(os.path.join(header_path, header)).read().decode('utf8')
-    build.append(re_depends.sub(lambda x:'\n', d))
+    d = open(pt.join(header_path, header)).read().decode('utf8')
+    build.append(re_depends.sub(lambda x: '\n', d))
     build.append('\n')
 
-open(OUTPUT,'w').write('\n'.join(build))
-
+open(OUTPUT, 'w').write('\n'.join(build))

--- a/amalgamate/merge_all.py
+++ b/amalgamate/merge_all.py
@@ -15,7 +15,7 @@ headers = [x.rsplit('/', 1)[-1] for x in glob(pt.join(header_path, '*.h'))]
 print(headers)
 edges = defaultdict(list)
 for header in headers:
-    d = open(pt.join(header_path, header)).read().decode('utf8')
+    d = open(pt.join(header_path, header)).read()
     match = re_depends.findall(d)
     for m in match:
         # m should included before header
@@ -47,7 +47,7 @@ for x in edges:
 print(order)
 build = []
 for header in order:
-    d = open(pt.join(header_path, header)).read().decode('utf8')
+    d = open(pt.join(header_path, header)).read()
     build.append(re_depends.sub(lambda x: '\n', d))
     build.append('\n')
 

